finality-aleph/src/justification/mod.rs

```diff
@@ -113,73 +102,22 @@ where
+        metrics: Option<Metrics<<B::Header as Header>::Hash>>,
+        verifier_timeout: Duration,
+        notification_timeout: Duration,
     ) -> Self {
         Self {
             session_info_provider,
-            block_requester,
-            client,
-            finalizer,
-            config,
-            phantom: PhantomData,
-        }
-    }
-
-    fn handle_justification_notification(
-        &mut self,
-        notification: JustificationNotification<B>,
-        verifier: V,
-        last_finalized: NumberFor<B>,
-        stop_h: NumberFor<B>,
-    ) {
-        let JustificationNotification {
-            justification,
-            number,
-            hash,
-        } = notification;
-
-        if number <= last_finalized || number > stop_h {
-            debug!(target: "afa", "Not finalizing block {:?}. Last finalized {:?}, stop_h {:?}", number, last_finalized, stop_h);
-            return;
-        };
-
-        if !(verifier.verify(&justification, hash)) {
-            warn!(target: "afa", "Error when verifying justification for block {:?} {:?}", number, hash);
-            return;
-        };
-
-        debug!(target: "afa", "Finalizing block {:?} {:?}", number, hash);
-        let finalization_res = self.finalizer.finalize_block(
-            hash,
-            number,
-            Some((ALEPH_ENGINE_ID, justification.encode())),
-        );
-        match finalization_res {
-            Ok(()) => {
-                self.config.justification_request_delay.on_block_finalized();
-                debug!(target: "afa", "Successfully finalized {:?}", number);
-                if let Some(metrics) = &self.config.metrics {
-                    metrics.report_block(hash, Instant::now(), Checkpoint::Finalized);
-                }
-            }
-            Err(e) => {
-                error!(target: "afa", "Fail in finalization of {:?} {:?} -- {:?}", number, hash, e);
-            }
-        }
-    }
-
-    fn request_justification(&mut self, num: NumberFor<B>) {
-        if self.config.justification_request_delay.can_request_now() {
-            debug!(target: "afa", "Trying to request block {:?}", num);
-
-            if let Ok(Some(header)) = self.client.header(BlockId::Number(num)) {
-                debug!(target: "afa", "We have block {:?} with hash {:?}. Requesting justification.", num, header.hash());
-                self.config.justification_request_delay.on_request_sent();
-                self.block_requester
-                    .request_justification(&header.hash(), *header.number());
-            } else {
-                debug!(target: "afa", "Cancelling request, because we don't have block {:?}.", num);
-            }
+            block_requester: BlockRequester::new(
+                block_requester,
+                client,
+                finalizer,
+                justification_request_delay,
+                metrics,
+            ),
+            verifier_timeout,
+            notification_timeout,
```

Comment:
I think the `Config` struct might have been useful here after all, although not quite in the way it was used before. You could pass all the things that are now `const`ants (in addition to the already configurable things you pass now) and only make them `Default`s in an implementation. You would need to deconstruct the `Config` struct  in the constructor and pass some of them to the requester, and keep the rest.

In this setup I would only pass more or less numeric values in the config – `JustificationRequestDelay` and metrics should be passed as separate arguments.